edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,693 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
+{
+    using System;
+    using System.Net;
+    using System.Text;
+    using System.Threading.Tasks;
+    using System.Web;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class RegistryController : Controller
+    {
+        const string ContentType = "application/json; charset=utf-8";
+
+        readonly IRegistryOnBehalfOfApiClient apiClient;
+        readonly Task<IHttpRequestAuthenticator> authenticatorGetter;
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public RegistryController(
+            IRegistryOnBehalfOfApiClient apiClient,
+            Task<IEdgeHub> edgeHubGetter,
+            Task<IHttpRequestAuthenticator> authenticatorGetter)
+        {
+            this.apiClient = Preconditions.CheckNotNull(apiClient, nameof(apiClient));
+            this.authenticatorGetter = Preconditions.CheckNotNull(authenticatorGetter, nameof(authenticatorGetter));
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHubGetter, nameof(edgeHubGetter));
+        }
+
+        [HttpPut]
+        [Route("devices/{deviceId}/modules/{moduleId}")]
+        public async Task CreateOrUpdateModuleAsync(
+            [FromRoute] string deviceId,
+            [FromRoute] string moduleId,
+            [FromHeader(Name="if-match")] string ifMatchHeader,
+            [FromBody] Module module)
+        {
+            try
+            {
+                Events.ReceivedRequest(nameof(this.CreateOrUpdateModuleAsync), deviceId, moduleId);
+
+                try
+                {
+                    deviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId)));
+                    moduleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId)));
+                    Preconditions.CheckNotNull(module, nameof(module));
+
+                    if (!string.Equals(deviceId, module.DeviceId) || !string.Equals(moduleId, module.Id))
+                    {
+                        throw new ApplicationException("Device Id or module Id doesn't match between request URI and body.");
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(nameof(this.CreateOrUpdateModuleAsync), ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, FormatErrorResponseMessage(ex.Message));
+                    return;
+                }
+
+                if (!await this.AuthenticateAsync(deviceId, Option.None<string>()))
```

Comment:
Why do we pass none here for moduleId?

Reply:
As this is the endpoint, call by iotedged.  Therefore no module id for this codepath.

Reply:
It shouldn't hurt to pass in the module ID though, it would in theory allow custom modules to call this API as well.

---
In reply to: [490643207](https://github.com/Azure/iotedge/pull/3532#discussion_r490643207) [](ancestors = 490643207)

Reply:
We should not assume who is making the call here. 
Also, if this is always called by a device, why does it have module in the route?

Reply:
What I understand is, an edge device owns its module and create/update module identity using their credential.  Therefore device Id should be considered only.  Module Id in the route is used for module identity creation.